src/ufuncs.c

```diff
@@ -448,30 +545,129 @@ static void *union_data[1] = {GEOSUnion_r};
 static void *shared_paths_data[1] = {GEOSSharedPaths_r};
 typedef void *FuncGEOS_YY_Y(void *context, void *a, void *b);
 static char YY_Y_dtypes[3] = {NPY_OBJECT, NPY_OBJECT, NPY_OBJECT};
+
+/* There are two inner loop functions for the YY_Y. This is because this
+ * pattern allows for ufunc.reduce.
+ * A reduce operation requires special attention, because we output the
+ * result in a temporary array. NumPy expects that the output array is
+ * filled during the inner loop, so that it can take the first input of
+ * the reduction operation to be the output array. Easiest to show by
+ * example:
+
+ * function called:         out = intersection.reduce(in)
+ * initialization by numpy: out[0] = in[0]; in1 = out; in2[:] = in[:]
+ * first loop:              out[0] = func(in1[0], in2[1])  [ = func(out[0], in2[1]) ]
+ * second loop:             out[0] = func(in1[0], in2[2])  [ = func(out[0], in2[1]) ]
+ */
+static void YY_Y_func_reduce(char **args, npy_intp *dimensions,
+                             npy_intp *steps, void *data)
+{
+    FuncGEOS_YY_Y *func = (FuncGEOS_YY_Y *)data;
+    GEOSGeometry *in1 = NULL, *in2 = NULL, *out = NULL;
+
+    GEOS_INIT_THREADS;
+
+    if (!get_geom(*(GeometryObject **)args[0], &out)) {
+        errstate = PGERR_NOT_A_GEOMETRY;
+    } else if (out != NULL) {
+        BINARY_LOOP {
+            // Cleanup previous in1
+            if (i > 1) {
+                // If i == 0, in1 is undefined
+                // If i == 1, in1 is the first input, which is owned by python
+                GEOSGeom_destroy_r(ctx, in1);
+            }
+            // This is the main reduce logic: in1 becomes previous out
+            in1 = out;
+            // Get the other geometry (as normal)
+            if (!get_geom(*(GeometryObject **)ip2, &in2)) {
+                errstate = PGERR_NOT_A_GEOMETRY;
+                break;
+            }
+            if (in2 == NULL) {
+                // We break the loop as the answer will remain NULL (None)
+                out = NULL;
+                break;
+            } else {
+                out = func(ctx, in1, in2);
+                if (out == NULL) {
+                    errstate = PGERR_GEOS_EXCEPTION;
+                    break;
+                }
+            }
+        }
+        // We need to cleanup the previous in1, but not if it is owned by python
+        // Get the first in1 from the ufunc input again (misuse in2 for that) and compare
```

Comment:
I don't fully understand this comment. In what kind of cases is in1 owned by python? If it is the original input geometry?